vscode: extract fork logic to support dedicated fork modules

emaiax opened the pull request:
### Description

<!-- Please provide a brief description of your change. -->

based on https://github.com/nix-community/home-manager/issues/6355, all logic from `vscode/default.nix` was extracted and refactored into a unified `vscodeFork` function that can be reused to quickly build more VS Code compatible forks, inspired by https://github.com/nix-community/home-manager/pull/5684.

given the number of vscode flavors (codium, cursor, windsurf, kiro, etc.), this makes it very easy for anyone to store and manage their editor configurations through their dedicated modules (`programs.cursor`, `programs.vscodium`, `programs.windsurf`, etc.).

**changes:**

- split main `vscode/default.nix` into specialized submodules (`profiles/settings.nix`, `profiles/snippets.nix`, `profiles/extensions.nix`) for better organization
- all vscode forks now share the same api through `vscode/vscodeFork.nix`
- easy addition of new forks on `vscode/default.nix`
- introduced new `snippets` option split into `global / languages` to improve organization
- all currently supported forks continue to be supported, with the addition that they now have their own dedicated home-manager modules via `programs.cursor`, `programs.vscodium`, `programs.windsurf`, etc
- added extensive tests covering all profile options, migration paths, files mutability and **unknown vscode-forks**

**supported flavors:**

- `programs.cursor` - Cursor
- `programs.kiro` - Kiro
- `programs.openvscode-server` - OpenVSCode Server
- `programs.vscode-insiders` - Visual Studio Code Insiders
- `programs.vscode` - Visual Studio Code
- `programs.vscodium` - VSCodium
- `programs.windsurf` - Windsurf

this PR may automatically address issues as:

- #7726
- #7597 
- #6355 
- #6305
- #3671 
- #2388
- maybe others, I'll start to maintain this module

**examples:**

```nix
programs.vscode = {
  enable = true;

  profiles = {
    default = {
      settings = { "files.autoSave" = "on"; };
      extensions = [ pkgs.vscode-extensions.bbenoist.nix ];
    };

    work = {
      settings = { "git.autofetch" = true; };
      extensions = [ pkgs.vscode-extensions.github.copilot ];
    };
  };
};

# or use the forks directly
programs.cursor = {
  enable = true;

  profiles.default = {
    settings = { "files.autoSave" = "on"; };

    mcp = {
      servers = {
        github = {
          command = "npx";
          args = [ "-y" "@modelcontextprotocol/server-github" ];
        };
      };
    };
  };
};
```

### Checklist

<!--
Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at
  https://nix-community.github.io/home-manager/#sec-guidelines
-->

- [x] Change is backwards compatible.
- [x] Code formatted with `nix fmt` or `nix-shell -p treefmt nixfmt deadnix keep-sorted --run treefmt`.
- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix build --reference-lock-file flake.lock ./tests#test-all` using Flakes.
- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).
- [x] Commit messages are formatted like
  ```
  {component}: {description}

  {long description}
  ```
See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.
- If this PR adds a new module
  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/a51598236f23c89e59ee77eb8e0614358b0e896c/modules/programs/lesspipe.nix#L11).

khaneliman commented:
Would highly recommend squashing into fewer commits and rebasing on latest master. Each commit should be an atomic operation that can eval/function properly by itself and documented behavioral changes in the commit body.

emaiax commented:
> Would highly recommend squashing into fewer commits and rebasing on latest master. Each commit should be an atomic operation that can eval/function properly by itself and documented behavioral changes in the commit body.

First of all, thank you for such a quick reply on a Saturday, I really appreciate the commitment of the other side! 👏🇧🇷 

Now, I rebased on top of latest master and made it compatible with commit guidelines. I think it should be ready for review and maybe trigger CI without caching.

I'll keep an eye on review comments. Thanks a lot! :)

emaiax commented:
Ok, sadly the CI still fails with cached past failures. Let's see what maintainers decides. Thank you!

emaiax committed:
vscode: refactor module into unified fork system

Extract all logic from `vscode/default.nix` into a unified `vscodeFork` function that enables easy creation of VS Code compatible fork modules. Split the main module into specialized submodules (profiles, snippets, and extensions) for better organization.
All forks now share the same API through `vscodeFork.nix` and are available through dedicated modules (`programs.cursor`, `programs.vscodium`, `programs.windsurf`, etc.). Backwards compatibility is maintained through automatic migrations from old API versions and a few bugs squashed in the process.
Comprehensive tests were added covering profiles, migration paths, file mutability, and unknown vscode-forks.

emaiax commented:
hey @khaneliman! 👋
I wonder if I'm missing anything on this PR.

khaneliman commented:
> hey @khaneliman! 👋 I wonder if I'm missing anything on this PR.

Looks like we buildbot failures that need to be addressed. Can you try rebasing your PR ?  I've been super busy end of year stuff so haven't had time for big PR reviews atm.

khaneliman commented on modules/programs/vscode/profiles/extensions.nix:
```diff
@@ -0,0 +1,166 @@
+  ];
+
+  getDefaultProfile = getAttrKey "default" cfg.profiles;
+  hasDefaultProfile = getDefaultProfile != null;
```
Move closer to usage or inline if only single usage.

khaneliman commented on modules/programs/vscode/profiles/extensions.nix:
```diff
@@ -0,0 +1,166 @@
+  hasDefaultProfile = getDefaultProfile != null;
+
+  getOtherProfiles = lib.removeAttrs cfg.profiles [ "default" ];
+  hasOtherProfiles = getOtherProfiles != { };
```
Same as above.

khaneliman commented on modules/programs/vscode/profiles/extensions.nix:
```diff
@@ -0,0 +1,166 @@
+  # - symlinks from the immutable extension directory to the nix store extension directory
+  # - extensions can be updated because it downloads a new version into the mutable extensions directory
+  #
+  buildExtensionsPaths =
```
^

khaneliman reviewed (commented):
Got distracted from functional changes with just very much style differences from the rest of our repo. Would like to avoid usage of rec, scope variables closer to usage, and move `lib` style module to a lib naming/organization. We have other examples in repo.
